File: Python/algorithms/graph_algorithms/a_star.py
"""
Algorithm: A* (A-star) Search
Description: Finds the shortest path between a start node and a goal node using 
             g(n) (actual cost so far) + h(n) (heuristic).
Time Complexity: O(E) in worst case
Space Complexity: O(V)
Author: Kashyap
Date: 2025-10-04
"""

import heapq

class Node:
    def __init__(self, position, parent=None, g=0, h=0):
        self.position = position  # (row, col)
        self.parent = parent      # previous node in path
        self.g = g                # cost from start
        self.h = h                # heuristic (to goal)
        self.f = g + h            # total cost

    def __lt__(self, other):
        return self.f < other.f   # needed for heapq priority

def heuristic(a, b):
    """Manhattan distance heuristic"""
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

def a_star(grid, start, goal):
    """
    Perform A* search on a grid.
    
    Args:
        grid (list[list[int]]): 0 = free cell, 1 = obstacle
        start (tuple): (row, col) start position
        goal (tuple): (row, col) goal position
    
    Returns:
        path (list[tuple]): shortest path from start to goal
    """
    open_set = []
    heapq.heappush(open_set, Node(start, None, 0, heuristic(start, goal)))
    visited = set()

    while open_set:
        current = heapq.heappop(open_set)

        if current.position == goal:
            # Reconstruct path
            path = []
            while current:
                path.append(current.position)
                current = current.parent
            return path[::-1]  # reverse path

        visited.add(current.position)

        # Explore neighbors (up, down, left, right)
        directions = [(0,1), (0,-1), (1,0), (-1,0)]
        for d in directions:
            neighbor = (current.position[0] + d[0], current.position[1] + d[1])

            # Skip if out of bounds or obstacle
            if (0 <= neighbor[0] < len(grid) and
                0 <= neighbor[1] < len(grid[0]) and
                grid[neighbor[0]][neighbor[1]] == 0 and
                neighbor not in visited):

                g_cost = current.g + 1
                h_cost = heuristic(neighbor, goal)
                heapq.heappush(open_set, Node(neighbor, current, g_cost, h_cost))
    
    return None  # no path found


# main func()
if __name__ == "__main__":
    # Test Case 1: Normal case
    grid1 = [
        [0, 1, 0, 0, 0],
        [0, 1, 0, 1, 0],
        [0, 0, 0, 1, 0],
        [0, 1, 0, 0, 0],
        [0, 0, 0, 1, 0]
    ]
    start1 = (0, 0)
    goal1 = (4, 4)
    path1 = a_star(grid1, start1, goal1)
    print("Test Case 1 Path:", path1) # Expected: a valid path like [(0,0),(1,0),(2,0),(2,1),...(4,4)]

    # Test Case 2: Blocked path
    grid2 = [
        [0, 1, 1],
        [1, 1, 1],
        [0, 0, 0]
    ]
    start2 = (0, 0)
    goal2 = (2, 2)
    path2 = a_star(grid2, start2, goal2)
    print("Test Case 2 Path:", path2) # Expected: None (no path exists)

    # Test Case 3: Start = Goal
    grid3 = [
        [0, 0],
        [0, 0]
    ]
    start3 = (0, 0)
    goal3 = (0, 0)
    path3 = a_star(grid3, start3, goal3)
    print("Test Case 3 Path:", path3) # Expected: [(0, 0)] (trivial path)

    # Test Case 4: Larger grid with obstacles
    grid4 = [
        [0, 0, 0, 0, 0, 0],
        [1, 1, 1, 1, 1, 0],
        [0, 0, 0, 0, 0, 0],
        [0, 1, 1, 1, 1, 1],
        [0, 0, 0, 0, 0, 0]
    ]
    start4 = (0, 0)
    goal4 = (4, 5)
    path4 = a_star(grid4, start4, goal4)
    print("Test Case 4 Path:", path4) # Expected: a valid path from top-left to bottom-right

File: CPP/data_structures/graphs/graph/AStar.cpp
#include <bits/stdc++.h>
using namespace std;

// Structure to represent a node in the grid
struct Node {
    int x, y;       // Coordinates
    int gCost;      // Cost from start node
    int hCost;      // Heuristic cost to goal
    int fCost;      // Total cost = g + h
    Node* parent;   // Pointer to parent node for path reconstruction

    Node(int _x, int _y, int _g, int _h, Node* _parent = nullptr)
        : x(_x), y(_y), gCost(_g), hCost(_h), parent(_parent) {
        fCost = gCost + hCost;
    }
};

// Heuristic function: Manhattan distance
int heuristic(int x1, int y1, int x2, int y2) {
    return abs(x1 - x2) + abs(y1 - y2);
}

// Check if a position is valid and walkable
bool isValid(int x, int y, int rows, int cols, vector<vector<int>>& grid) {
    return x >= 0 && y >= 0 && x < rows && y < cols && grid[x][y] == 0;
}

// A* algorithm to find shortest path
vector<pair<int,int>> aStar(vector<vector<int>>& grid, pair<int,int> start, pair<int,int> goal) {
    int rows = grid.size();
    int cols = grid[0].size();

    auto cmp = [](Node* a, Node* b) { return a->fCost > b->fCost; };
    priority_queue<Node*, vector<Node*>, decltype(cmp)> openSet(cmp);

    vector<vector<bool>> closedSet(rows, vector<bool>(cols, false));

    Node* startNode = new Node(start.first, start.second, 0, heuristic(start.first, start.second, goal.first, goal.second));
    openSet.push(startNode);

    int dx[] = {-1, 1, 0, 0};
    int dy[] = {0, 0, -1, 1};

    while (!openSet.empty()) {
        Node* current = openSet.top();
        openSet.pop();

        if (current->x == goal.first && current->y == goal.second) {
            // Reconstruct path
            vector<pair<int,int>> path;
            while (current) {
                path.push_back({current->x, current->y});
                current = current->parent;
            }
            reverse(path.begin(), path.end());
            return path;
        }

        if (closedSet[current->x][current->y]) continue;
        closedSet[current->x][current->y] = true;

        for (int i = 0; i < 4; i++) {
            int nx = current->x + dx[i];
            int ny = current->y + dy[i];

            if (isValid(nx, ny, rows, cols, grid) && !closedSet[nx][ny]) {
                int g = current->gCost + 1;
                int h = heuristic(nx, ny, goal.first, goal.second);
                Node* neighbor = new Node(nx, ny, g, h, current);
                openSet.push(neighbor);
            }
        }
    }

    return {}; // No path found
}

// Example usage
int main() {
    vector<vector<int>> grid = {
        {0, 1, 0, 0, 0},
        {0, 1, 0, 1, 0},
        {0, 0, 0, 1, 0},
        {0, 1, 1, 0, 0},
        {0, 0, 0, 0, 0}
    };

    pair<int,int> start = {0, 0};
    pair<int,int> goal = {4, 4};

    vector<pair<int,int>> path = aStar(grid, start, goal);

    if (!path.empty()) {
        cout << "Path found:\n";
        for (auto p : path) {
            cout << "(" << p.first << "," << p.second << ") ";
        }
        cout << endl;
    } else {
        cout << "No path found.\n";
    }

    return 0;
}
